Npm install is no re-run if installation fails during dependency extraction

To reproduce:
1. Download a starter
2. Run `mvn`  once to generate all needed files
3. `rm -rf node_modules`
4. `npm install` and ctrl-c when it says `extract:something`
5. Run `mvn`

The output is 
```
2020-04-01 17:27:22.375  INFO 40069 --- [  restartedMain] c.v.f.s.frontend.TaskUpdatePackages      : Skipping `npm install`.
2020-04-01 17:27:22.375  INFO 40069 --- [  restartedMain] c.v.f.s.frontend.TaskCopyFrontendFiles   : Copying frontend resources from jar files ...
2020-04-01 17:27:22.413  INFO 40069 --- [  restartedMain] c.v.f.s.frontend.TaskCopyFrontendFiles   : Visited 13 resources. Took 38 ms.
...
  Failed to find the following imports in the `node_modules` tree:
      - @vaadin/vaadin-details/src/vaadin-details.js
      - @vaadin/vaadin-grid/src/vaadin-grid-column.js
      - @vaadin/vaadin-ordered-layout/src/vaadin-vertical-layout.js
      - @vaadin/vaadin-lumo-styles/spacing.js
      - @vaadin/vaadin-item/src/vaadin-item.js
      - @vaadin/vaadin-text-field/src/vaadin-number-field.js
      - @vaadin/vaadin-lumo-styles/badge.js
      - @vaadin/vaadin-icons/vaadin-icons.js
      - @vaadin/flow-frontend/ironListStyles.js
      - @vaadin/vaadin-upload/src/vaadin-upload.js
      - @vaadin/vaadin-dialog/src/vaadin-dialog.js
      - @vaadin/vaadin-select/src/vaadin-select.js
      - @vaadin/flow-frontend/vaadin-big-decimal-field.js
      - @vaadin/vaadin-button/src/vaadin-button.js
      - @vaadin/flow-frontend/contextMenuConnector.js
      - @vaadin/flow-frontend/dndConnector-es6.js
      - @vaadin/vaadin-date-picker/src/vaadin-date-picker.js
      - @vaadin/vaadin-menu-bar/src/vaadin-menu-bar.js
      - @vaadin/vaadin-grid-pro/src/vaadin-grid-pro-edit-column.js
      - @vaadin/vaadin-time-picker/src/vaadin-time-picker.js
      - @vaadin/vaadin-combo-box/src/vaadin-combo-box.js
      - @vaadin/vaadin-context-menu/src/vaadin-context-menu.js
      - @vaadin/vaadin-grid-pro/src/vaadin-grid-pro.js
      - @vaadin/vaadin-tabs/src/vaadin-tabs.js
      - @vaadin/vaadin-cookie-consent/src/vaadin-cookie-consent.js
      - @vaadin/vaadin-list-box/src/vaadin-list-box.js
      - @vaadin/vaadin-text-field/src/vaadin-text-field.js
      - @vaadin/vaadin-app-layout/src/vaadin-drawer-toggle.js
      - @vaadin/vaadin-notification/src/vaadin-notification.js
      - @vaadin/vaadin-radio-button/src/vaadin-radio-button.js
      - @vaadin/vaadin-split-layout/src/vaadin-split-layout.js
      - @vaadin/vaadin-accordion/src/vaadin-accordion.js
      - @vaadin/flow-frontend/gridConnector.js
      - @vaadin/vaadin-login/src/vaadin-login-overlay.js
      - @polymer/iron-list/iron-list.js
      - @vaadin/vaadin-radio-button/src/vaadin-radio-group.js
      - @polymer/iron-icon/iron-icon.js
      - @vaadin/vaadin-upload/src/vaadin-upload-file.js
      - @vaadin/flow-frontend/comboBoxConnector.js
      - @vaadin/vaadin-form-layout/src/vaadin-form-item.js
      - @vaadin/vaadin-tabs/src/vaadin-tab.js
      - @vaadin/vaadin-grid/src/vaadin-grid-column-group.js
      - @vaadin/flow-frontend/timepickerConnector.js
      - @vaadin/vaadin-crud/src/vaadin-crud-edit-column.js
      - @vaadin/flow-frontend/selectConnector.js
      - @vaadin/vaadin-confirm-dialog/src/vaadin-confirm-dialog.js
      - @vaadin/vaadin-lumo-styles/typography.js
      - @vaadin/vaadin-app-layout/src/vaadin-app-layout.js
      - @vaadin/vaadin-checkbox/src/vaadin-checkbox-group.js
      - @vaadin/vaadin-custom-field/src/vaadin-custom-field.js
      - @vaadin/vaadin-lumo-styles/sizing.js
      - @vaadin/vaadin-checkbox/src/vaadin-checkbox.js
      - @vaadin/flow-frontend/menubarConnector.js
      - @vaadin/flow-frontend/vaadin-grid-flow-selection-column.js
      - @vaadin/flow-frontend/datepickerConnector.js
      - @vaadin/vaadin-charts/vaadin-chart.js
      - @vaadin/vaadin-text-field/src/vaadin-password-field.js
      - @vaadin/vaadin-board/vaadin-board.js
      - @vaadin/flow-frontend/ironListConnector.js
      - @vaadin/vaadin-ordered-layout/src/vaadin-horizontal-layout.js
      - @vaadin/vaadin-text-field/src/vaadin-text-area.js
      - @vaadin/vaadin-login/src/vaadin-login-form.js
      - @vaadin/vaadin-board/vaadin-board-row.js
      - @vaadin/vaadin-text-field/src/vaadin-integer-field.js
      - @vaadin/vaadin-grid/src/vaadin-grid-sorter.js
      - @vaadin/vaadin-progress-bar/src/vaadin-progress-bar.js
      - @vaadin/vaadin-rich-text-editor/src/vaadin-rich-text-editor.js
      - @vaadin/vaadin-lumo-styles/icons.js
      - @vaadin/flow-frontend/gridProConnector.js
      - @vaadin/vaadin-grid/src/vaadin-grid-tree-toggle.js
      - @vaadin/vaadin-crud/src/vaadin-crud.js
      - @vaadin/vaadin-lumo-styles/style.js
      - @vaadin/vaadin-grid/src/vaadin-grid.js
      - @vaadin/flow-frontend/flow-component-renderer.js
      - @vaadin/vaadin-lumo-styles/color.js
      - @vaadin/vaadin-form-layout/src/vaadin-form-layout.js
      - @vaadin/vaadin-text-field/src/vaadin-email-field.js
  If the build fails, check that npm packages are installed.

  To fix the build remove `pnpm-lock.yaml` and `node_modules` directory to reset modules.
  In addition you may run `pnpm install` to fix `node_modules` tree structure.
Make sure first that `pnpm` command is installed, otherwise you should install it using npm: `npm add -g pnpm@4.5.0`


2020-04-01 17:27:22.451  INFO 40069 --- [  restartedMain] c.v.f.server.frontend.TaskUpdateImports  : Updated /.../frontend/generated-flow-imports.js
2020-04-01 17:27:22.452  INFO 40069 --- [  restartedMain] c.v.f.server.frontend.TaskUpdateImports  : No js modules to update '.../target/frontend/generated-flow-imports.d.ts' file
2020-04-01 17:27:22.454  WARN 40069 --- [  restartedMain] com.vaadin.flow.server.DevModeHandler    : Instance not created because cannot execute '.../node_modules/webpack-dev-server/bin/webpack-dev-server.js'. Did you run `npm install`
2020-04-01 17:27:22.612  INFO 40069 --- [  restartedMain] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'applicationTaskExecutor'
```


The problems are:
1. `npm install` is not run as `node_modules` contains a `.staging` folder
2. The build contines even though A LOT of critical dependencies are missing
3. `webpack-dev-server` fails to start yet the build is allowed to continue
4. The end of the build output is
```
2020-04-01 17:29:37.826  INFO 40120 --- [  restartedMain] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
2020-04-01 17:29:37.831  INFO 40120 --- [  restartedMain] com.example.application.Application      : Started Application in 5.331 seconds (JVM running for 5.789)
```
which indicates everything is fine
5. Opening the app in the browser only shows
```
Whitelabel Error Page
This application has no explicit mapping for /error, so you are seeing this as a fallback.

Wed Apr 01 17:30:03 EEST 2020
There was an unexpected error (type=Internal Server Error, status=500).
java.lang.NullPointerException
com.vaadin.flow.server.ServiceException: java.lang.NullPointerException
	at com.vaadin.flow.server.VaadinService.handleExceptionDuringRequest(VaadinService.java:1588)
...
Caused by: java.lang.NullPointerException
	at com.vaadin.flow.server.frontend.FrontendUtils.getFileFromWebpack(FrontendUtils.java:778)
	at com.vaadin.flow.server.frontend.FrontendUtils.getFileContent(FrontendUtils.java:638)
	at com.vaadin.flow.server.frontend.FrontendUtils.getIndexHtmlContent(FrontendUtils.java:627)
```
with no information about the actual problem

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
@@ -51,7 +51,7 @@ public class TaskRunNpmInstall implements FallibleCommand {
     private final NodeUpdater packageUpdater;
 
     private final List<String> ignoredNodeFolders = Arrays.asList(".bin",
-            "pnpm", ".ignored_pnpm", ".pnpm", MODULES_YAML);
+            "pnpm", ".ignored_pnpm", ".pnpm", ".staging", MODULES_YAML);
     private final boolean enablePnpm;
     private final boolean requireHomeNodeExec;
     private final ClassFinder classFinder;

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
@@ -90,6 +90,18 @@ public class TaskRunNpmInstallTest {
 
         Mockito.verify(logger).info(getRunningMsg());
     }
+    @Test
+    public void runNpmInstall_nodeModulesContainsStaging_npmInstallIsExecuted()
+            throws ExecutionFailedException {
+        File nodeModules = getNodeUpdater().nodeModulesFolder;
+        nodeModules.mkdir();
+        File staging = new File(nodeModules, ".staging");
+        staging.mkdir();
+        nodeUpdater.modified = false;
+        task.execute();
+
+        Mockito.verify(logger).info(getRunningMsg());
+    }
 
     @Test
     public void runNpmInstall_toolIsChanged_nodeModulesIsRemoved()

